README.md

```diff
@@ -7,6 +7,7 @@ An Ansible role for installing [Apache Spark](https://spark.apache.org).
 - `spark_version` - Spark version.
 - `spark_cloudera_distribution` - Cloudera distribution version (default: `cdh5.4`)
 - `spark_env_extras` - An optional dictionary with key and value attributes to add to `spark-env.sh` (e.g. `MESOS_NATIVE_LIBRARY: "/usr/local/lib/libmesos.so"`)
+- `spark_defaults_extras` - An optional list (not a dictionary, to keep given ordering) with settings to `spark-defaults.conf` (e.g. `"spark.eventLog.enabled   true"`)
```

Comment:
We've discussed this internally and prefer an approach that uses a `dict` for `spark_defaults_extras` and then applies a `sort` before templating to ensure a consistent order (although not a natural order). I personally have not seen a scenario where Ansible rearranged the order of items in a `dict` when iterating, but if it is possible, the `sort` should account for that.

Here is the patch we used to confirm these approaches:

```diff
diff --git a/README.md b/README.md
index b7036f7..c2d75ea 100644
--- a/README.md
+++ b/README.md
@@ -7,7 +7,8 @@ An Ansible role for installing [Apache Spark](https://spark.apache.org).
 - `spark_version` - Spark version.
 - `spark_cloudera_distribution` - Cloudera distribution version (default: `cdh5.4`)
 - `spark_env_extras` - An optional dictionary with key and value attributes to add to `spark-env.sh` (e.g. `MESOS_NATIVE_LIBRARY: "/usr/local/lib/libmesos.so"`)
-- `spark_defaults_extras` - An optional list (not a dictionary, to keep given ordering) with settings to `spark-defaults.conf` (e.g. `"spark.eventLog.enabled   true"`)
+- `spark_defaults_extras` - An optional dictionary with key and value attributes
+  to add to `spark-defaults.conf` (e.g. `"spark.eventLog.enabled": true`)
 
 ## Example Playbook
 
diff --git a/defaults/main.yml b/defaults/main.yml
index 615f955..bbcc065 100644
--- a/defaults/main.yml
+++ b/defaults/main.yml
@@ -12,4 +12,4 @@ spark_user: "spark"           # the name of the (OS)user created for spark
 spark_user_groups: []         # Optional list of (OS)groups the new spark user should belong to
 
 spark_env_extras: {}
-spark_defaults_extras: []
+spark_defaults_extras: {}
diff --git a/examples/site.yml b/examples/site.yml
index 12825d5..0016d20 100644
--- a/examples/site.yml
+++ b/examples/site.yml
@@ -3,6 +3,12 @@
 
   vars:
     java_version: "7u51-2.4.*"
+    spark_env_extras:
+      TEST_B: "b"
+      TEST_A: "a"
+    spark_defaults_extras:
+      "spark.eventLog.enabled": "true"
+      "spark.eventLog.disabled": "true"
 
   pre_tasks:
     - name: Update APT cache
diff --git a/templates/spark-defaults.conf.j2 b/templates/spark-defaults.conf.j2
index f3e261e..8c50a54 100644
--- a/templates/spark-defaults.conf.j2
+++ b/templates/spark-defaults.conf.j2
@@ -9,6 +9,6 @@
 # spark.driver.memory              5g
 # spark.executor.extraJavaOptions  -XX:+PrintGCDetails -Dkey=value -Dnumbers="one two three"
 
-{% for setting in spark_defaults_extras %}
-{{ setting }}
+{% for key, value in spark_defaults_extras.items() | sort %}
+{{ key }} {{ value }}
 {% endfor %}
diff --git a/templates/spark-env.sh.j2 b/templates/spark-env.sh.j2
index 4ab1172..45f2d43 100644
--- a/templates/spark-env.sh.j2
+++ b/templates/spark-env.sh.j2
@@ -56,6 +56,6 @@ export HADOOP_CONF_DIR=${HADOOP_CONF_DIR:-/etc/hadoop/conf}
 export SPARK_LOG_DIR=${SPARK_LOG_DIR:-{{ spark_log_dir }}}
 export SPARK_PID_DIR=${SPARK_PID_DIR:-{{ spark_run_dir }}}
 
-{% for key, value in spark_env_extras.items() %}
+{% for key, value in spark_env_extras.items() | sort %}
 export {{ key }}="{{ value }}"
 {% endfor %}
```

Does that seem acceptable?